src/Xamarin.Android.Build.Tasks/Utilities/MarshalMethodsAssemblyRewriter.cs

```diff
@@ -0,0 +1,147 @@
+		}
+
+		public void Rewrite (DirectoryAssemblyResolver resolver)
+		{
+			MethodDefinition unmanagedCallersOnlyAttributeCtor = GetUnmanagedCallersOnlyAttributeConstructor (resolver);
+			var unmanagedCallersOnlyAttributes = new Dictionary<AssemblyDefinition, CustomAttribute> ();
+			foreach (AssemblyDefinition asm in uniqueAssemblies) {
+				unmanagedCallersOnlyAttributes.Add (asm, CreateImportedUnmanagedCallersOnlyAttribute (asm, unmanagedCallersOnlyAttributeCtor));
+			}
+
+			Console.WriteLine ("Adding the [UnmanagedCallersOnly] attribute to native callback methods and removing unneeded fields+methods");
+			foreach (IList<MarshalMethodEntry> methodList in methods.Values) {
+				foreach (MarshalMethodEntry method in methodList) {
+					Console.WriteLine ($"\t{method.NativeCallback.FullName} (token: 0x{method.NativeCallback.MetadataToken.RID:x})");
+					method.NativeCallback.CustomAttributes.Add (unmanagedCallersOnlyAttributes [method.NativeCallback.Module.Assembly]);
+					method.Connector.DeclaringType.Methods.Remove (method.Connector);
+					method.CallbackField?.DeclaringType.Fields.Remove (method.CallbackField);
+				}
+			}
+
+			Console.WriteLine ();
+			Console.WriteLine ("Rewriting assemblies");
+
+			var newAssemblyPaths = new List<string> ();
+			foreach (AssemblyDefinition asm in uniqueAssemblies) {
+				foreach (string path in GetAssemblyPaths (asm)) {
+					var writerParams = new WriterParameters {
+						WriteSymbols = (File.Exists (path + ".mdb") || File.Exists (Path.ChangeExtension (path, ".pdb"))),
+					};
+
+					string output = $"{path}.new";
+					Console.WriteLine ($"\t{asm.Name} => {output}");
+					asm.Write (output, writerParams);
+					newAssemblyPaths.Add (output);
+				}
+			}
+
+			// Replace old versions of the assemblies only after we've finished rewriting without issues, otherwise leave the new
+			// versions around.
+			foreach (string path in newAssemblyPaths) {
+				string target = Path.Combine (Path.GetDirectoryName (path), Path.GetFileNameWithoutExtension (path));
+				MoveFile (path, target);
+
+				string source = Path.ChangeExtension (path, ".pdb");
+				if (File.Exists (source)) {
+					target = Path.ChangeExtension (Path.Combine (Path.GetDirectoryName (source), Path.GetFileNameWithoutExtension (source)), ".pdb");
+
+					MoveFile (source, target);
+				}
+
+				source = $"{path}.mdb";
+				if (File.Exists (source)) {
+					target = Path.ChangeExtension (path, ".mdb");
+					MoveFile (source, target);
+				}
+			}
+
+			Console.WriteLine ();
+			Console.WriteLine ("Method tokens:");
+			foreach (IList<MarshalMethodEntry> methodList in methods.Values) {
+				foreach (MarshalMethodEntry method in methodList) {
+					Console.WriteLine ($"\t{method.NativeCallback.FullName} (token: 0x{method.NativeCallback.MetadataToken.RID:x})");
+				}
+			}
+
+			void MoveFile (string source, string target)
+			{
+				Console.WriteLine ($"Moving '{source}' => '{target}'");
```

Comment:
We have a CopyIfChanged methd in `Files` which will only copy a file over if there are changes. It does few other things like making sure the file is writable etc. It might not be as quick, but it does handle a bunch of issue which we have hit before.

Reply:
That said it won't delete the source file, which I this does.

Reply:
Changed to use `CopyIfChanged` and `File.Delete`